Polyhedron/demo/Polyhedron/Scene_polyhedron_item.cpp

```diff
@@ -41,31 +41,164 @@
-            item->setProperty(aabb_property_name,
-                              QVariant::fromValue<void*>(tree));
-            return tree;
+    previous = vh;
+  } while( ++he_circ != he_circ_end );
+  cdt.insert_constraint(previous, first);
+  // sets mark is_external
+  for(CDT::All_faces_iterator
+      fit2 = cdt.all_faces_begin(),
+      end = cdt.all_faces_end();
+      fit2 != end; ++fit2)
+  {
+    fit2->info().is_external = false;
+  }
+  //check if the facet is external or internal
+  std::queue<CDT::Face_handle> face_queue;
+  face_queue.push(cdt.infinite_vertex()->face());
+  while(! face_queue.empty() ) {
+    CDT::Face_handle fh = face_queue.front();
+    face_queue.pop();
+    if(fh->info().is_external) continue;
+    fh->info().is_external = true;
+    for(int i = 0; i <3; ++i) {
+      if(!cdt.is_constrained(std::make_pair(fh, i)))
+      {
+        face_queue.push(fh->neighbor(i));
+      }
+    }
+  }
+  //iterates on the internal faces to add the vertices to the positions
+  //and the normals to the appropriate vectors
+  for(CDT::Finite_faces_iterator
+      ffit = cdt.finite_faces_begin(),
+      end = cdt.finite_faces_end();
+      ffit != end; ++ffit)
+  {
+    if(ffit->info().is_external)
+      continue;
+
+
+    res << Kernel::Triangle_3(ffit->vertex(0)->point(),
+    ffit->vertex(1)->point(),
+    ffit->vertex(2)->point());
+
+  }
+  return res;
+}
+
+
+Input_facets_AABB_tree* get_aabb_tree(Scene_polyhedron_item* item)
+{
+  QVariant aabb_tree_property = item->property(aabb_property_name);
+  if(aabb_tree_property.isValid()) {
+    void* ptr = aabb_tree_property.value<void*>();
+    return static_cast<Input_facets_AABB_tree*>(ptr);
+  }
+  else {
+    Polyhedron* poly = item->polyhedron();
+    if(poly) {
+      //creates an empty AABB_Tree
+      Input_facets_AABB_tree* tree =
+          new Input_facets_AABB_tree()/*faces(*poly).first,
+                                                               faces(*poly).second,
+                                                               *poly)*/;
+      typedef Polyhedron::Traits	    Kernel;
+      typedef Kernel::Vector_3	    Vector;
+      typedef boost::graph_traits<Polyhedron>::face_descriptor   face_descriptor;
+      typedef boost::graph_traits<Polyhedron>::vertex_descriptor vertex_descriptor;
+
+      boost::container::flat_map<face_descriptor, Vector> face_normals_map;
+      boost::associative_property_map< boost::container::flat_map<face_descriptor, Vector> >
+        nf_pmap(face_normals_map);
+      boost::container::flat_map<vertex_descriptor, Vector> vertex_normals_map;
+      boost::associative_property_map< boost::container::flat_map<vertex_descriptor, Vector> >
+        nv_pmap(vertex_normals_map);
+
+      PMP::compute_normals(*poly, nv_pmap, nf_pmap);
+      Q_FOREACH( Polyhedron::Facet_iterator f, faces(*poly))
+      {
+        //triangulates the primitives
+        triangulate_primitive(f,nf_pmap);
```

Comment:
if the mesh is already triangled, there's no need to compute face and vertex normals.